specs/backlog/node-owner-policy.rst

```diff
@@ -0,0 +1,195 @@
+ License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+
+======================================
+Adding Policy Targets for Ironic Nodes
+======================================
+
+Include the URL of your StoryBoard story (which should have an `rfe`` tag):
+
+https://storyboard.openstack.org/#!/story/XXXXXXX
+
+This spec describes an update to the Ironic node controller to expose Owner
+node data for policy checks.
+
+
+Problem description
+===================
+
+Right now Ironic is not multi-tenant; anyone with API access to one node
+has access to all nodes. Nodes have an ``owner`` field; however it is
+purely informational and not tied into any form of access control. This
+is inconvenient if an Ironic deployment consists of nodes from multiple
+hardware Owners: each Owner needs to be able to manage their own nodes
+through the Ironic API, while being unable to manage others' nodes.
+
+
+Proposed change
+===============
+
+The Ironic API already has a way of controlling access to the REST API: a
+policy engine[0] that is used throughout the REST API. However, when the node
+controller uses the policy engine[1], it does so without passing in any
+information about the node being checked. Other services such as Nova[2] and
+Barbican[3] pass in additional information, and we propose doing the same. In
+summary, we would like to:
+
+* Update the node controller so that policy checks also pass in information
+  about the node, including ``owner``
+* Update Ironic's default generated policy file to include ``is_node_owner``
+  rules, along with updated rules for ``node:*`` actions (Nova does the same[4]).
+  For example:
+
+   *  "is_node_owner": "project_id:%(node.owner)s"
+   *  "baremetal:node:set_power_state": "rule:is_admin or rule:is_node_owner"
+
+This update is enough to control access for most API functions - except for list
+functions. For those, we propose updating the node list query so that non-admin
+results are filtered against the ``owner`` field.
+
+Alternatives
+------------
+
+One alternative is to perform these checks at the database API level. However
+that requires a new code pattern to be used on a large number of individual
+functions.
+
+Data model impact
+-----------------
+
+None
+
+State Machine Impact
+--------------------
+
+None
+
+REST API impact
+---------------
+
+None
+
+Client (CLI) impact
+-------------------
+
+None
+
+"ironic" CLI
+~~~~~~~~~~~~
+
+None
+
+"openstack baremetal" CLI
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+None
+
+RPC API impact
+--------------
+
+None
+
+Driver API impact
+-----------------
+
+None
+
+Nova driver impact
+------------------
+
+None
+
+Ramdisk impact
+--------------
+
+None
+
+Security impact
+---------------
+
+This change exposes functionality to additional users. However this access
+is constrained through Oslo policy, and can be adjusted as an adminstrator
+desires.
+
+Other end user impact
+---------------------
+
+None
```

Comment:
I think that when our policy engine is active, we should validate the value of the `owner` field when it is set or updated. Would this count as a REST API change?

Reply:
Does that mean we should have a configuration setting so we know when it's active?

I'm also not sure it's necessary from an OpenStack-pattern point of view.